# Chapter 4 Problem 10

In [97]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.metrics import confusion_matrix

In [98]:
weekly = pd.read_csv("Weekly.csv")

In [99]:
weekly.head()

Unnamed: 0  Year   Lag1   Lag2   Lag3   Lag4   Lag5    Volume  Today  \
0           1  1990  0.816  1.572 -3.936 -0.229 -3.484  0.154976 -0.270   
1           2  1990 -0.270  0.816  1.572 -3.936 -0.229  0.148574 -2.576   
2           3  1990 -2.576 -0.270  0.816  1.572 -3.936  0.159837  3.514   
3           4  1990  3.514 -2.576 -0.270  0.816  1.572  0.161630  0.712   
4           5  1990  0.712  3.514 -2.576 -0.270  0.816  0.153728  1.178   

  Direction  
0      Down  
1      Down  
2        Up  
3        Up  
4        Up

In [100]:
weekly.shape

(1089, 10)

** NEED TO DO NUMERICAL/GRAPHICAL SUMMARIES **

## Logistic Regression on all data

In [101]:
logReg = LogisticRegression()
testing = weekly.loc[:,"Lag1":"Volume"]
testing.head()

Lag1   Lag2   Lag3   Lag4   Lag5    Volume
0  0.816  1.572 -3.936 -0.229 -3.484  0.154976
1 -0.270  0.816  1.572 -3.936 -0.229  0.148574
2 -2.576 -0.270  0.816  1.572 -3.936  0.159837
3  3.514 -2.576 -0.270  0.816  1.572  0.161630
4  0.712  3.514 -2.576 -0.270  0.816  0.153728

In [102]:
logReg.fit(testing, weekly["Direction"])

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [103]:
testing.describe()

Lag1         Lag2         Lag3         Lag4         Lag5  \
count  1089.000000  1089.000000  1089.000000  1089.000000  1089.000000   
mean      0.150585     0.151079     0.147205     0.145818     0.139893   
std       2.357013     2.357254     2.360502     2.360279     2.361285   
min     -18.195000   -18.195000   -18.195000   -18.195000   -18.195000   
25%      -1.154000    -1.154000    -1.158000    -1.158000    -1.166000   
50%       0.241000     0.241000     0.241000     0.238000     0.234000   
75%       1.405000     1.409000     1.409000     1.409000     1.405000   
max      12.026000    12.026000    12.026000    12.026000    12.026000   

            Volume  
count  1089.000000  
mean      1.574618  
std       1.686636  
min       0.087465  
25%       0.332022  
50%       1.002680  
75%       2.053727  
max       9.328214

The matrix shows the result of describe() on the matrix made up of the predictor variables. All 5 Lag variables are very similar to each other. The Volume variable, by its nature, has a different structure, but just based on the above matrix, it is hard to say whether any of these variables are by themselves statistically significant.

## Confusion Matrix

In [104]:
results = logReg.predict(testing)
print(results.shape)
results

(1089,)


array(['Up', 'Up', 'Up', ..., 'Up', 'Up', 'Up'], dtype=object)

In [105]:
print(confusion_matrix(weekly["Direction"], results))

[[ 55 429]
 [ 47 558]]


Figuring out which row corresponds to "Up" and which row corresponds to "Down" - 

In [106]:
up_correct = 0
up_predicted_as_down = 0
for num in range(1089):
    if (weekly["Direction"][num]=="Up" and results[num]=="Up"):
        up_correct+=1
    if (weekly["Direction"][num]=="Up" and results[num]=="Down"):
        up_predicted_as_down +=1
print(up_correct)
print(up_predicted_as_down)

558
47


Interpretation of the confusion matrix - 
55 Downs and 558 Ups are predicted correctly. 47 Ups are predicted as downs and 429 Downs are predicted as ups.

Overall fraction of correct predictions - (55+558)/ (47+429+55+558) = 0.563

## Logistic Regression tested on 1990-2008 data

In [107]:
pre_2009 = weekly.loc[:984, :]
pre_2009.tail()

Unnamed: 0  Year    Lag1    Lag2    Lag3    Lag4    Lag5    Volume  \
980         981  2008  12.026  -8.389  -6.198  -3.898  10.491  5.841565   
981         982  2008  -2.251  12.026  -8.389  -6.198  -3.898  6.093950   
982         983  2008   0.418  -2.251  12.026  -8.389  -6.198  5.932454   
983         984  2008   0.926   0.418  -2.251  12.026  -8.389  5.855972   
984         985  2008  -1.698   0.926   0.418  -2.251  12.026  3.087105   

     Today Direction  
980 -2.251      Down  
981  0.418        Up  
982  0.926        Up  
983 -1.698      Down  
984  6.760        Up

This doesn't seem a great way to separate the pre-2009 data from the rest but for now, this is how I will do it. I will change this when I figure out how to do this a better way

In [108]:
post_2009 = weekly.loc[985:, :]
post_2009.head()

Unnamed: 0  Year   Lag1   Lag2   Lag3   Lag4   Lag5    Volume  Today  \
985         986  2009  6.760 -1.698  0.926  0.418 -2.251  3.793110 -4.448   
986         987  2009 -4.448  6.760 -1.698  0.926  0.418  5.043904 -4.518   
987         988  2009 -4.518 -4.448  6.760 -1.698  0.926  5.948758 -2.137   
988         989  2009 -2.137 -4.518 -4.448  6.760 -1.698  6.129763 -0.730   
989         990  2009 -0.730 -2.137 -4.518 -4.448  6.760  5.602004  5.173   

    Direction  
985      Down  
986      Down  
987      Down  
988      Down  
989        Up

In [109]:
lag2_training  = pre_2009["Lag2"]
lag2_training.shape

(985,)

In [110]:
lag2_training = np.reshape(lag2_training, (985, 1)) # to get rid of a deprecation warning
new_logistic_regr = LogisticRegression()
new_logistic_regr.fit(lag2_training, pre_2009["Direction"])

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [111]:
lag2_testing = post_2009["Lag2"]
lag2_testing.shape

(104,)

In [112]:
lag2_testing = np.reshape(lag2_testing, (104, 1))
new_results = new_logistic_regr.predict(lag2_testing)
print(confusion_matrix(post_2009["Direction"], new_results))

[[ 9 34]
 [ 5 56]]


Assuming the same row-wise interpretation of the confusion matrix as before, the interpretation of the matrix is as follows - 
9 Downs and 56 Ups are predicted correctly. 5 Ups are predicted as Downs and 34 Downs are predicted as Ups.
Overall fraction of correct predictions = (9+34)/(9+34+5+56) = 0.413

## Linear Discriminant Analysis

In [113]:
lag2_training.shape

(985, 1)

In [114]:
lda = LDA()
lda.fit(lag2_training, pre_2009["Direction"])

LinearDiscriminantAnalysis(n_components=None, priors=None, shrinkage=None,
              solver='svd', store_covariance=False, tol=0.0001)

In [115]:
lda_results = lda.predict(lag2_testing)

In [116]:
print(confusion_matrix(post_2009["Direction"], lda_results))

[[ 9 34]
 [ 5 56]]


This is the exact same confusion matrix as the last logistic regression model. Let's see if both the models have the same predictions for each instance.

In [119]:
x1 = new_logistic_regr.predict_log_proba(lag2_training)
x2 = lda.predict_log_proba(lag2_training)
for num in range(5): #printing out only the first few
    print(x1[num], x2[num])

[-0.85074208 -0.55704038] [-0.85097261 -0.55686856]
[-0.82583009 -0.57602457] [-0.82621748 -0.57572292]
[-0.79087438 -0.60412607] [-0.79147622 -0.60362703]
[-0.719923   -0.66706965] [-0.72093726 -0.66610856]
[-0.91688866 -0.5104272 ] [-0.91668814 -0.51056077]


As the log probabilities returned by both these models are not the same, the fact that both these models return the same confusion matrix is not an issue.

Interpretration of the confusion matrix - 
9 Downs and 56 Ups are predicted correctly. 5 Ups are predicted as Downs and 34 Downs are predicted as Ups. Overall fraction of correct predictions = (9+34)/(9+34+5+56) = 0.413